src/workerd/jsg/async-context.c++

```diff
@@ -0,0 +1,297 @@
+      args.GetReturnValue().Set(result);
+    }
+  }, fn));
+}
+
+v8::Local<v8::Promise> AsyncContextFrame::attachContext(
+    Lock& js,
+    v8::Local<v8::Promise> promise,
+    kj::Maybe<AsyncContextFrame&> maybeFrame) {
+  auto handle = js.getPrivateSymbolFor("asyncResource"_kj);
+  auto context = js.v8Isolate->GetCurrentContext();
+  // If the promise has already been wrapped, do nothing else and just return the promise.
+  if (!check(promise->HasPrivate(context, handle))) {
+    // Otherwise, we have to create an opaque wrapper holding a ref to the current frame
+    // because we do not have the option of using an internal field with promises.
+    auto frame = kj::addRef(AsyncContextFrame::current(js));
+    KJ_ASSERT(check(promise->SetPrivate(context, handle, wrapOpaque(context, kj::mv(frame)))));
+  }
+  return promise;
+}
+
+kj::Maybe<Value&> AsyncContextFrame::get(StorageKey& key) {
+  KJ_ASSERT(!key.isDead());
+  storage.eraseAll([](const auto& entry) { return entry.key->isDead(); });
+  return storage.find(key).map([](auto& entry) -> Value& { return entry.value; });
+}
+
+AsyncContextFrame::Scope::Scope(Lock& js, kj::Maybe<AsyncContextFrame&> resource)
+    : Scope(js.v8Isolate, resource) {}
+
+AsyncContextFrame::Scope::Scope(v8::Isolate* isolate, kj::Maybe<AsyncContextFrame&> frame)
+    : isolate(IsolateBase::from(isolate)) {
+  KJ_IF_MAYBE(f, frame) {
+    this->isolate.pushAsyncFrame(*f);
+  } else {
+    this->isolate.pushAsyncFrame(*this->isolate.rootAsyncFrame);
+  }
+}
+
+AsyncContextFrame::Scope::~Scope() noexcept(false) {
+  isolate.popAsyncFrame();
+}
+
+AsyncContextFrame::StorageScope::StorageScope(
+    Lock& js,
+    StorageKey& key,
+    Value store)
+    : frame(AsyncContextFrame::create(js, nullptr, StorageEntry {
+        .key = kj::addRef(key),
+        .value = kj::mv(store)
+      })),
+      scope(js, *frame) {}
+
+bool AsyncContextFrame::isRoot(Lock& js) const {
+  return IsolateBase::from(js.v8Isolate).rootAsyncFrame == this;
+}
+
+void IsolateBase::pushAsyncFrame(AsyncContextFrame& next) {
+  asyncFrameStack.push_front(&next);
+}
+
+void IsolateBase::popAsyncFrame() {
+  asyncFrameStack.pop_front();
+  KJ_DASSERT(!asyncFrameStack.empty(), "the async context frame stack was corrupted");
+}
+
+void IsolateBase::promiseHook(v8::PromiseHookType type,
+                              v8::Local<v8::Promise> promise,
+                              v8::Local<v8::Value> parent) {
+  auto isolate = promise->GetIsolate();
+
+  // V8 will call the promise hook even while execution is terminating. In that
+  // case we don't want to do anything here.
+  if (isolate->IsExecutionTerminating() || isolate->IsDead()) {
+    return;
+  }
+
+  auto& js = Lock::from(isolate);
+  auto& isolateBase = IsolateBase::from(isolate);
+  auto& currentFrame = AsyncContextFrame::current(js);
+
+  const auto isRejected = [&] { return promise->State() == v8::Promise::PromiseState::kRejected; };
+
+  js.tryCatch([&] {
```

Comment:
`js.tryCatch()` explicitly doesn't catch terminations, which could be generated during the callback since it might do allocation (and therefore exceed the memory limit).

We need something more like `liftKj`, but that currently requires a `FunctionCallbackInfo`. Maybe we need a new variant of `liftKj` for callbacks that don't have callback info?

Reply:
> Maybe we need a new variant of liftKj for callbacks that don't have callback info?

+1